tools/deploy/helper.go

```diff
@@ -0,0 +1,427 @@
+	toAPIEndpointFn    ToAPIEndpointFn
+	placementWatcher   services.StagedPlacementWatcher
+	settleBetweenSteps time.Duration
+}
+
+// NewHelper creates a new deployment helper.
+func NewHelper(opts HelperOptions) (Helper, error) {
+	client := newAggregatorClient(opts.HTTPClient())
+	planner := newPlanner(client, opts.PlannerOptions())
+	placementWatcherOpts := opts.StagedPlacementWatcherOptions()
+	placementWatcher := placement.NewStagedPlacementWatcher(placementWatcherOpts)
+	if err := placementWatcher.Watch(); err != nil {
+		return nil, err
+	}
+	retryOpts := opts.RetryOptions()
+	retrier := xretry.NewRetrier(retryOpts)
+	foreverRetrier := xretry.NewRetrier(retryOpts.SetForever(true))
+	return helper{
+		logger:             opts.InstrumentOptions().Logger(),
+		planner:            planner,
+		client:             client,
+		mgr:                opts.Manager(),
+		store:              opts.KVStore(),
+		retrier:            retrier,
+		foreverRetrier:     foreverRetrier,
+		workers:            opts.WorkerPool(),
+		toPlacementIDFn:    opts.ToPlacementInstanceIDFn(),
+		toAPIEndpointFn:    opts.ToAPIEndpointFn(),
+		placementWatcher:   placementWatcher,
+		settleBetweenSteps: opts.SettleDurationBetweenSteps(),
+	}, nil
+}
+
+func (h helper) Deploy(revision string, mode Mode) error {
+	if revision == "" {
+		return errInvalidRevision
+	}
+	all, err := h.allInstanceMetadatas()
+	if err != nil {
+		return fmt.Errorf("unable to get all instance metadatas: %v", err)
+	}
+	filtered := filterByRevision(all, revision)
+
+	plan, err := h.planner.GeneratePlan(filtered, all)
+	if err != nil {
+		return fmt.Errorf("unable to generate deployment plan: %v", err)
+	}
+
+	// If in dry run mode, log the generated deployment plan and return.
+	if mode == DryRunMode {
+		h.logger.Infof("Generated deployment plan: %+v", plan)
+		return nil
+	}
+
+	if err = h.execute(plan, revision, all); err != nil {
+		return fmt.Errorf("unable to execute deployment plan: %v", err)
+	}
+
+	return nil
+}
+
+// Close closes the deployment helper.
+func (h helper) Close() error {
+	if h.placementWatcher == nil {
+		return nil
+	}
+	return h.placementWatcher.Unwatch()
+}
+
+func (h helper) execute(
+	plan deploymentPlan,
+	revision string,
+	all instanceMetadatas,
+) error {
+	numSteps := len(plan.Steps)
+	for i, step := range plan.Steps {
+		h.logger.Infof("deploying step %d of %d", i+1, numSteps)
+		if err := h.executeStep(step, revision, all); err != nil {
+			return err
+		}
+		h.logger.Infof("deploying step %d succeeded", i+1)
+		if i < numSteps-1 && h.settleBetweenSteps > 0 {
+			h.logger.Infof("waiting settle duration after step: %s", h.settleBetweenSteps.String())
+			time.Sleep(h.settleBetweenSteps)
+		}
+	}
+	return nil
+}
+
+func (h helper) executeStep(
+	step deploymentStep,
+	revision string,
+	all instanceMetadatas,
+) error {
+	h.logger.Infof("waiting until safe to deploy for step %v", step)
+	if err := h.waitUntilSafe(all); err != nil {
```

Comment:
Yeah so the rationale is to essentially wait for all the agg tier nodes to be healthy and not actively deployed to before initiating another deploy. This is primarily for extra caution in case  there are some nodes that have gone down after they are newly deployed, in which case we probably want to investigate before proceeding with a new round of deploy.